File: C++/2-Medium/00318._最大单词长度乘积/00318._最大单词长度乘积.cpp
#include <vector>
#include <string>
#include <iostream>
using namespace std;

class Solution {
public:
    int maxProduct(vector<string>& words) {
        int n = words.size();
        vector<int> masks(n, 0); // 用于存储每个字符串的字符掩码
        int maxProduct = 0;
        
        // 构建每个字符串的掩码
        for (int i = 0; i < n; ++i) {
            for(char c: words[i]){
                masks[i] |= 1 << (c - 'a');
            }
        }
        
        // 比较每对字符串
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if ((masks[i] & masks[j]) == 0) { // 检查是否没有共同字符
                    maxProduct = max(maxProduct, int(words[i].length() * words[j].length()));
                }
            }
        }
        
        return maxProduct;
    }
};

int main() {
    Solution solution;
    vector<string> words = {"abcw","baz","foo","bar","xtfn","abcdef"};
    cout << "Maximum product of word lengths: " << solution.maxProduct(words) << endl;
    return 0;
}

File: Python/2-Medium/00318._最大单词长度乘积/00318._最大单词长度乘积.py
from typing import List

def maxProduct(words: List[str]) -> int:
    """
    计算不包含相同字符的两个单词长度乘积的最大值。
    
    Args:
        words: 字符串列表，每个元素是一个单词。
    
    Returns:
        两个单词长度乘积的最大值。
        
    示例:
        >>> maxProduct(["abcw","baz","foo","bar","xtfn","abcdef"])
        16
    """
    # 将单词转换为二进制表示，并记录其长度
    word_bits = {}  # key为单词的二进制表示，value为单词的长度
    for word in words:
        bits = 0
        for char in word:
            # 计算字符在字母表中的位置，并设置相应的位
            bits |= 1 << (ord(char) - ord('a'))
        # 记录单词的最大长度（相同二进制表示可能对应不同长度的单词）
        word_bits[bits] = max(word_bits.get(bits, 0), len(word))
    
    max_product = 0
    # 比较所有单词，寻找最大长度乘积
    for x in word_bits:
        for y in word_bits:
            if x & y == 0:  # 如果没有公共字母
                max_product = max(max_product, word_bits[x] * word_bits[y])
                
    return max_product

# 测试代码
test_cases = [
    ["abcw","baz","foo","bar","xtfn","abcdef"],
    ["a","ab","abc","d","cd","bcd","abcd"],
    ["a","aa","aaa","aaaa"]
]

results = [maxProduct(words) for words in test_cases]
print(results)